server/src/snapshot.rs

```
@@ -0,0 +1,307 @@
+}
+
+impl Status {
+    fn next_table_position(&mut self) -> Option<usize> {
+        self.remaining_tables.pop().map(|pos| {
+            self.running_tables.push(pos);
+            pos
+        })
+    }
+
+    fn mark_table_finished(&mut self, position: usize) -> Result<()> {
+        let pos = self
+            .running_tables
+            .iter()
+            .position(|t| *t == position)
+            .context(TableNotFound { table: position })?;
+
+        let _ = self.running_tables.remove(pos);
+        self.finished_tables.push(position);
+
+        Ok(())
+    }
+
+    fn finished(&self) -> bool {
+        self.remaining_tables.is_empty() && self.running_tables.is_empty() && self.meta_written
+    }
+}
+
+pub fn snapshot_partition<T: Send + Sync + 'static + Partition>(
+    metadata_path: impl Into<String>,
+    data_path: impl Into<String>,
+    store: Arc<ObjectStore>,
+    partition: Arc<T>,
+    notify: Option<oneshot::Sender<()>>,
+) -> Result<Arc<Snapshot>> {
+    let metadata_path = metadata_path.into();
+    let data_path = data_path.into();
+
+    let table_stats = partition.table_stats().unwrap();
+
+    let snapshot = Snapshot::new(partition.key().to_string(), table_stats);
+    let snapshot = Arc::new(snapshot);
+
+    let return_snapshot = snapshot.clone();
+
+    tokio::spawn(async move {
+        while let Some((pos, table_name)) = snapshot.next_table() {
+            let batch = partition.table_to_arrow(table_name, &[]).unwrap();
+
+            let file_name = format!("{}/{}.parquet", &data_path, table_name);
+
+            write_batch(batch, snapshot.clone(), &file_name, store.clone())
+                .await
+                .unwrap();
+
+            snapshot.mark_table_finished(pos).unwrap();
+        }
+
+        let partition_meta_path = format!("{}/{}.json", &metadata_path, &partition.key());
+        let json_data = serde_json::to_vec(&snapshot.partition_meta)
+            .context(JsonGenerationError)
+            .unwrap();
```

Since this runs with a separate task, my experience with gRPC implementation is that it is critical to make sure these errors get logged somehow (I know you said you are going to work on this prior to merge). 

One pattern I was doing, that bit me, was use `?` here but not doing anything with the return value of `tokio_spawn`. In that case, the error that was returned by `?` got silently dropped which caused me non trivial confusion / debugging pain

---

I ended up logging the error via `log_if_error` -- https://github.com/influxdata/influxdb_iox/blob/main/data_types/src/error.rs#L24 in addition to trying to look at the result

---

Added the error logging in d7e8c3243cc4636fe7f9ec1419b3ff8c71fbc5ee. I ended up refactoring things a bit so the snapshot holds everything. I think it turned out a bit cleaner and may be helpful in the future if we wire up the API to pull information from the partition chunk a snapshot is working with.